File: app/src/main/java/com/buddha/mindboard/util/CustomViewModelFactory.java
package com.buddha.mindboard.util;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.support.annotation.NonNull;

import com.buddha.mindboard.data.repository.Repository;
import com.buddha.mindboard.module.home.HomeViewModel;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class CustomViewModelFactory implements ViewModelProvider.Factory {
    private final Repository repository;

    @Inject
    public CustomViewModelFactory(Repository repository) {
        this.repository = repository;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(HomeViewModel.class))
            return (T) new HomeViewModel(repository);
        else
            throw new IllegalArgumentException("ViewModel not found!");
    }
}

File: app/src/main/java/com/buddha/mindboard/util/CustomViewModelFactory.kt
package com.buddha.mindboard.util

import android.arch.lifecycle.ViewModel
import android.arch.lifecycle.ViewModelProvider

import com.buddha.mindboard.data.repository.Repository
import com.buddha.mindboard.module.home.HomeViewModel

import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class CustomViewModelFactory @Inject
constructor(private val repository: Repository) : ViewModelProvider.Factory {

    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        return if (modelClass.isAssignableFrom(HomeViewModel::class.java))
            HomeViewModel(repository) as T
        else
            throw IllegalArgumentException("ViewModel not found!")
    }
}
